File: Go/SwapNodesInPairs.go
//go:build ignore

package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	nxt := head.Next
	head.Next = swapPairs(nxt.Next)
	nxt.Next = head
	return nxt
}

func build(vals []int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range vals {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return dummy.Next
}

func toSlice(h *ListNode) []int {
	out := []int{}
	for h != nil {
		out = append(out, h.Val)
		h = h.Next
	}
	return out
}

func main() {
	fmt.Println(toSlice(swapPairs(build([]int{1, 2, 3, 4})))) // [2 1 4 3]
	fmt.Println(toSlice(swapPairs(build([]int{}))))           // []
	fmt.Println(toSlice(swapPairs(build([]int{1}))))          // [1]
	fmt.Println(toSlice(swapPairs(build([]int{1, 2, 3}))))    // [2 1 3]
}

File: CPP/SwapNodesInPairs.cpp
#include <initializer_list>
#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v = 0, ListNode* n = nullptr) : val(v), next(n) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head || !head->next) return head;
        ListNode* nxt = head->next;
        head->next = swapPairs(nxt->next);
        nxt->next = head;
        return nxt;
    }
};

ListNode* build(std::initializer_list<int> vals) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vals) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

void print(ListNode* h) {
    std::cout << "[";
    bool first = true;
    while (h) {
        if (!first) std::cout << ", ";
        std::cout << h->val;
        first = false;
        h = h->next;
    }
    std::cout << "]\n";
}

void freeList(ListNode* h) {
    while (h) {
        ListNode* n = h->next;
        delete h;
        h = n;
    }
}

int main() {
    Solution s;
    for (auto vals : std::vector<std::vector<int>>{ {1, 2, 3, 4}, {}, {1}, {1, 2, 3} }) {
        ListNode dummy;
        ListNode* cur = &dummy;
        for (int v : vals) { cur->next = new ListNode(v); cur = cur->next; }
        ListNode* result = s.swapPairs(dummy.next);
        print(result);
        freeList(result);
    }
    return 0;
}
